ingestor-default/src/main/scala/za/co/absa/hyperdrive/ingestor/implementation/utils/KafkaUtil.scala

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2018 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.hyperdrive.ingestor.implementation.utils
+
+import java.time.Duration
+
+import org.apache.kafka.clients.consumer.{ConsumerRecord, KafkaConsumer}
+import org.apache.kafka.common.TopicPartition
+import org.apache.logging.log4j.LogManager
+import org.apache.spark.sql.execution.streaming.{CommitLog, Offset, OffsetSeqLog}
+import org.apache.spark.sql.kafka010.KafkaSourceOffsetProxy
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+private[hyperdrive] object KafkaUtil {
+  private val logger = LogManager.getLogger
+
+  def getAtLeastNLatestRecordsFromPartition[K, V](consumer: KafkaConsumer[K, V], topicPartition: TopicPartition, numberOfRecords: Int)
+    (implicit kafkaConsumerTimeout: Duration): Seq[ConsumerRecord[K, V]] = {
+    consumer.assign(Seq(topicPartition).asJava)
+    val endOffsets = consumer.endOffsets(Seq(topicPartition).asJava).asScala
+    if (endOffsets.size != 1) {
+      throw new IllegalStateException(s"Expected exactly 1 end offset, got ${endOffsets}")
+    }
+    val partition = endOffsets.keys.head
+    val offset = endOffsets.values.head
+
+    var records: Seq[ConsumerRecord[K, V]] = Seq()
+    var offsetLowerBound = offset
+    while (records.size < numberOfRecords && offsetLowerBound != 0) {
```

Comment:
Maybe I don't see it here, but do you have a reason why you do this in the cycle? 
For example: 
Your endOffset is 100
numberOfRecords 10
In the first cycle you set consumer pointer to endOffset - numberOfRecords = 90 
do the reading from topic  getMessagesAtLeastToOffset(consumer, Map(topicPartition -> offset)) which says give me everything from 90 to offset (offset is endOffset = 100). This function gives you that or it throws error.

Reply:
The result is that you read everything what is on that topic

Reply:
At the end, you read all messages in topic

Reply:
Excellent question! If I remove the while-loop, the test `"getAtLeastNLatestRecords" should "get at least the n latest records"` in `TestKafkaUtilDockerTest` will fail on line 230 "6 was not greater than or equal 10" Only 6 messages are read, not 10 as expected. Why? This test case creates a compacted topic. A compacted topic is one of the cases where you can have gaps in the offsets.
So you cannot predict for sure how many records you are going to read. The only way to know how many records you got, is to consume and count them.

To answer your question, yes `getMessagesAtLeastToOffset` gives me everything from 90 to 100, but these aren't necessarily 10 messages. In this method however, I want 10 messages, not less. So I'm going to consume until I have 10 messages. I don't think I'm reading all the messages in the topic, since `numberOfRecords` is constant

Reply:
Nice!